### DISCUSSION TITLE
여담) 무지성 Feature에 관한 Discussion

### DISCUSSION THREAD

**Discussion by MoonJaeHoon** [bcaitech1/p4-dkt-freshtomato#49] [General]
## 무지성 Feature 관련해서 정리하는 글
(계속해서 업데이트될듯)
그냥 마구잡이로 생각없이 추가한 변수들이 현재 많은 상황인데, 성능 혹은 변수중요도를 체크해서 필요없는 변수들을 정리하려고 한다.
변수가 많아지다 보니 무의미한 변수도 조금 끼어있는 것 같네.. Feature Selection 과정을 날잡고 한번 거쳐줘야겠다!
> 맨 아래의 그림은 어느 LGBM의 Feature Importance Plot이다.

---
### 1. 문제풀이간격 변수(굉장히 의외의 결과)
의외? : **해당 변수를 그냥 추가하고 후처리도 아예 안해준 경우가 가장 성능이 좋았기 때문...**

이 변수를 핀트가 어긋나지 않게끔 제대로 만들기 위해서는 여러 케이스와 에외들을 고려해야 하고,
그에 따라 변수생성을 하기 위해서 정말 많은 애를 쓰고 있는데,
난 그냥 무작정 userID, testID 별로 groupby 한 다음 문제풀이 간의 시간간격을 구해 변수를 생성해줘보았다.
- 이렇게 생성하면 같은 시험지를 여러번 푼 학생에 한하여 이전의 마지막 문제 <-> 이후의 첫 문제 사이의 풀이 시간 간격이 매우 커서 문제가 생길 것으로 판단되었다.
- 그래서 3600초(=60분=1시간)를 기준으로 삼아 넘어가면 이는 0으로 고려해버리는 후처리를 진행하였다.
> 결론적으로 해당 변수가 있을 떄와 없을 때, 후처리를 하지 않았을 때의 결과를 다음과 같이 비교해보았는데 굉장히 의외였다.

Name|LB_ACC|LB_AUC
-- | -- | --
5-fold_optuna_all_train_userid_time_diff_shift-1_lgbm_trial50_문제풀이간격후처리X | 0.6290 | 0.7585
5-fold_optuna_all_train_userid_lgbm_trial50_문제풀이간격변수없음 | 0.6371 | 0.7566
5-fold_optuna_all_train_userid_time_diff_shift-1_lgbm_trial50_문제풀이간격후처리O | 0.6317 | 0.7559

---
### 2. KnowledgeTag 변수 무참히 슬라이싱 (민용이가 영감을 주었음)

아래의 결과에서 볼 수 있듯이 어떤 경우에는, 본래 KnowledgeTag를 쓰는 것보다 오히려 앞에서 3글자, 뒤에서 3글자를 고려한 변수가 더 비중있다고 평가되기도 했다.
하지만, 그 외에 자잘하게 슬라이싱하여 추가한 변수들은 본래보다 그리 중요한 변수로서 고려되지는 않는 모습이다.

---
### 3. 시간 관련 변수들
연도, 요일, 월, 몇번째 주인지, 계절변수, 시간대 등
(몇번째 주인지)week_number가 의외로 유의해보임
(풀이 시간대)hour는 예상대로 유의해보임
의외로 요일과 같은 변수가 유의하지 않아보임.

![image](https://user-images.githubusercontent.com/54096561/120071521-c2db2280-c0ca-11eb-8655-c0a601345224.png)
![image](https://user-images.githubusercontent.com/54096561/120071546-dedec400-c0ca-11eb-8426-612657824ca7.png)

**Comment by arabae** [comment]
몇번째 주인지는 시간이 지날수록 유저의 실력이 향상되어서 그런거일까?

**Reply by MoonJaeHoon** [reply]
아 이 변수가..
만약 개인 userID 별로 시작부터 현재까지 **몇 주가 지났는지 누적카운팅**해준 변수라면 그럴 수 있겠는데..
사실 지금 만들어놓은 변수는 해당 시점 정보인 TimeStamp 로부터 그냥 1년 중 몇번째 주인지 추출한 변수일 뿐이라서 어떤 의미를 찾기가 쫌 어렵네 ㅠㅠ
